pyvista/utilities/reader.py

```diff
@@ -186,8 +183,26 @@ def reader(self):
             raise NotImplementedError
         return self._reader
 
-    def _set_filename_or_directory(self, path):
-        """Set filename or directory and update reader."""
+    @property
+    def path(self) -> str:
+        """Return or set the filename or directory of the reader.
+
+        Examples
+        --------
+        >>> import pyvista
+        >>> from pyvista import examples
+        >>> filename = examples.download_human(load=False)
+        >>> reader = pyvista.XMLPolyDataReader(filename)
+        >>> reader.path  # doctest:+SKIP
+        '/home/user/.local/share/pyvista/examples/Human.vtp'
+
+        """
+        if self._filename is not None:
+            return self._filename
+        return self.__directory
+
+    @path.setter
+    def path(self, path: str):
         if os.path.isdir(path):
             self._set_directory(path)
         elif os.path.isfile(path):
```

Comment:
If you have an object where path starts as a filename and then is later set to be a directory, I think you will end up with both `_filename` and `__directory` having non-`None` values and the getter will return an inaccurate result (it will return the filename meanwhile the object is actually operating in directory mode).

The quick fix is to add some lines here to set `_filename` and `__directory` to `None` when the other one gets set in the path setter. Long term it would be less confusing if neither `_filename` nor `__directory` existed and `path` (or more approrpriate, `_path`) was used for everything and all of the children were updated to also use `path` or `_path` instead of `filename` or `_filename`.

Reply:
Resolved in 21a5d5d7f035ee0a5d33e3492d91d5a11e9c27bd